File: libs/headless-figma-clone/src/eval/heuristics.ts
import { findEnvelopeNode } from '../engine/DocumentEngine.js';
import type { AnyTreeNode, FileEnvelope, Paint, RGB, TextNode } from '../model/types.js';
import { collectContentNodeIds } from './designSystem.js';
import type { EditGraph, SubCheckResult } from './types.js';

const WCAG_AA_NORMAL = 4.5;
const MAX_DISTINCT_FONTS = 5;
const MIN_READABLE_FONT_SIZE = 10;

const DEFAULT_FG: RGB = { r: 0, g: 0, b: 0 };
const DEFAULT_BG: RGB = { r: 1, g: 1, b: 1 };

function heurResult(
  id: string,
  score: number,
  applicable: boolean,
  details?: Record<string, unknown>
): SubCheckResult {
  return { id, category: 'heuristics', score, applicable, weight: 1, details };
}

function firstSolidColor(paints: Paint[] | undefined): RGB | null {
  for (const paint of paints ?? []) {
    if (paint.type === 'SOLID') return paint.color;
  }
  return null;
}

function srgbChannel(c: number): number {
  return c <= 0.03928 ? c / 12.92 : ((c + 0.055) / 1.055) ** 2.4;
}

function relativeLuminance(color: RGB): number {
  const r = srgbChannel(color.r);
  const g = srgbChannel(color.g);
  const b = srgbChannel(color.b);
  return 0.2126 * r + 0.7152 * g + 0.0722 * b;
}

function contrastRatio(foreground: RGB, background: RGB): number {
  const l1 = relativeLuminance(foreground);
  const l2 = relativeLuminance(background);
  const lighter = Math.max(l1, l2);
  const darker = Math.min(l1, l2);
  return (lighter + 0.05) / (darker + 0.05);
}

function contrastScore(ratio: number): number {
  return Math.min(1, ratio / WCAG_AA_NORMAL);
}

function parentIdOf(envelope: FileEnvelope, nodeId: string): string | null {
  let found: string | null = null;
  const walk = (node: AnyTreeNode, parent: AnyTreeNode | null): void => {
    if (node.type === 'DOCUMENT') {
      for (const ch of node.children) walk(ch, node);
      return;
    }
    if (node.id === nodeId && parent && parent.type !== 'DOCUMENT') {
      found = parent.id;
      return;
    }
    if ('children' in node && Array.isArray((node as { children?: AnyTreeNode[] }).children)) {
      for (const ch of (node as { children: AnyTreeNode[] }).children) walk(ch, node);
    }
  };
  walk(envelope.document, null);
  return found;
}

function resolveBackgroundColor(envelope: FileEnvelope, nodeId: string): RGB {
  let current: string | null = parentIdOf(envelope, nodeId);
  while (current) {
    const node = findEnvelopeNode(envelope, current);
    if (node) {
      const rec = node as { fills?: Paint[]; backgrounds?: Paint[] };
      const fromFill = firstSolidColor(rec.fills) ?? firstSolidColor(rec.backgrounds);
      if (fromFill) return fromFill;
    }
    current = parentIdOf(envelope, current);
  }
  return DEFAULT_BG;
}

function textForegroundColor(node: TextNode): RGB {
  const fromFills = firstSolidColor(node.fills);
  if (fromFills) return fromFills;
  for (const seg of node.styledSegments ?? []) {
    const c = firstSolidColor(seg.style?.fills);
    if (c) return c;
  }
  return DEFAULT_FG;
}

function effectiveFontSize(node: TextNode, envelope: FileEnvelope): number | undefined {
  if (typeof node.fontSize === 'number') return node.fontSize;
  if (node.textStyleId) {
    const style = envelope.textStyles?.find((s) => s.id === node.textStyleId);
    if (typeof style?.fontSize === 'number') return style.fontSize;
  }
  for (const seg of node.styledSegments ?? []) {
    if (typeof seg.style?.fontSize === 'number') return seg.style.fontSize;
  }
  return undefined;
}

function fontSignature(node: TextNode, envelope: FileEnvelope): string {
  const size = effectiveFontSize(node, envelope);
  const styleName = node.textStyleId
    ? envelope.textStyles?.find((s) => s.id === node.textStyleId)?.name
    : undefined;
  const family = node.fontName?.family ?? styleName ?? 'default';
  return `${family}:${size ?? 'inherit'}`;
}

function textNodesInContent(
  after: FileEnvelope,
  contentIds: Set<string>
): TextNode[] {
  const nodes: TextNode[] = [];
  for (const id of contentIds) {
    const node = findEnvelopeNode(after, id);
    if (node?.type === 'TEXT') nodes.push(node);
  }
  return nodes;
}

function runContrast(
  after: FileEnvelope,
  contentIds: Set<string>
): SubCheckResult {
  const texts = textNodesInContent(after, contentIds);
  if (texts.length === 0) {
    return heurResult('heuristics.contrast', 1, false, { reason: 'no_text_in_content' });
  }

  const ratios: number[] = [];
  for (const text of texts) {
    const fg = textForegroundColor(text);
    const bg = resolveBackgroundColor(after, text.id);
    ratios.push(contrastRatio(fg, bg));
  }

  const meanRatio = ratios.reduce((a, b) => a + b, 0) / ratios.length;
  const score = contrastScore(meanRatio);

  return heurResult('heuristics.contrast', score, true, {
    texts_checked: texts.length,
    mean_contrast_ratio: meanRatio,
    min_ratio: Math.min(...ratios),
  });
}

function runFontCount(
  after: FileEnvelope,
  contentIds: Set<string>
): SubCheckResult {
  const texts = textNodesInContent(after, contentIds);
  if (texts.length === 0) {
    return heurResult('heuristics.font_count', 1, false, { reason: 'no_text_in_content' });
  }

  const signatures = new Set(texts.map((t) => fontSignature(t, after)));
  const count = signatures.size;
  const score =
    count <= MAX_DISTINCT_FONTS
      ? 1
      : Math.max(0, 1 - (count - MAX_DISTINCT_FONTS) / MAX_DISTINCT_FONTS);

  return heurResult('heuristics.font_count', score, true, {
    distinct_fonts: count,
    max_recommended: MAX_DISTINCT_FONTS,
  });
}

function runReadableFontSize(
  after: FileEnvelope,
  contentIds: Set<string>
): SubCheckResult {
  const texts = textNodesInContent(after, contentIds);
  if (texts.length === 0) {
    return heurResult('heuristics.readable_font_size', 1, false, { reason: 'no_text_in_content' });
  }

  let violations = 0;
  const sizes: number[] = [];
  for (const text of texts) {
    const size = effectiveFontSize(text, after);
    if (size === undefined) continue;
    sizes.push(size);
    if (size < MIN_READABLE_FONT_SIZE) violations += 1;
  }

  if (sizes.length === 0) {
    return heurResult('heuristics.readable_font_size', 1, false, { reason: 'no_resolved_font_sizes' });
  }

  const score = 1 - violations / sizes.length;
  return heurResult('heuristics.readable_font_size', score, true, {
    violations,
    sizes_checked: sizes.length,
    min_size: Math.min(...sizes),
  });
}

/** Low-weight quality heuristics on content-changed subtrees. */
export function runHeuristics(
  _before: FileEnvelope,
  after: FileEnvelope,
  graph: EditGraph
): SubCheckResult[] {
  const contentIds = collectContentNodeIds(graph, after);

  if (contentIds.size === 0) {
    return [
      heurResult('heuristics.contrast', 1, false, { reason: 'no_content_changes' }),
      heurResult('heuristics.font_count', 1, false, { reason: 'no_content_changes' }),
      heurResult('heuristics.readable_font_size', 1, false, { reason: 'no_content_changes' }),
    ];
  }

  return [
    runContrast(after, contentIds),
    runFontCount(after, contentIds),
    runReadableFontSize(after, contentIds),
  ];
}

File: envs/figma-design/tasks/with-metadata/tests/figma_eval/heuristics.py
from __future__ import annotations

from typing import Any

from .design_system import collect_content_node_ids
from .edit_graph import EditGraph
from .tree import find_node
from .types import Envelope, SubCheckResult, TreeNode

WCAG_AA_NORMAL = 4.5
MAX_DISTINCT_FONTS = 5
MIN_READABLE_FONT_SIZE = 10
DEFAULT_FG = {"r": 0.0, "g": 0.0, "b": 0.0}
DEFAULT_BG = {"r": 1.0, "g": 1.0, "b": 1.0}


def _heur_result(
    heur_id: str, score: float, applicable: bool, details: dict[str, Any] | None = None
) -> SubCheckResult:
    return SubCheckResult(
        id=heur_id,
        category="heuristics",
        score=score,
        applicable=applicable,
        weight=1.0,
        details=details,
    )


def _first_solid_color(paints: list[dict[str, Any]] | None) -> dict[str, float] | None:
    for paint in paints or []:
        if paint.get("type") == "SOLID":
            return paint["color"]
    return None


def _srgb_channel(c: float) -> float:
    return c / 12.92 if c <= 0.03928 else ((c + 0.055) / 1.055) ** 2.4


def _relative_luminance(color: dict[str, float]) -> float:
    r = _srgb_channel(color["r"])
    g = _srgb_channel(color["g"])
    b = _srgb_channel(color["b"])
    return 0.2126 * r + 0.7152 * g + 0.0722 * b


def _contrast_ratio(fg: dict[str, float], bg: dict[str, float]) -> float:
    l1, l2 = _relative_luminance(fg), _relative_luminance(bg)
    lighter, darker = max(l1, l2), min(l1, l2)
    return (lighter + 0.05) / (darker + 0.05)


def _parent_id_of(envelope: Envelope, node_id: str) -> str | None:
    found: str | None = None

    def walk(node: TreeNode, parent: TreeNode | None) -> None:
        nonlocal found
        if node.get("type") == "DOCUMENT":
            for ch in node.get("children") or []:
                walk(ch, node)
            return
        if node.get("id") == node_id and parent and parent.get("type") != "DOCUMENT":
            found = parent["id"]
            return
        for ch in node.get("children") or []:
            walk(ch, node)

    doc = envelope.get("document")
    if doc:
        walk(doc, None)
    return found


def _resolve_background_color(envelope: Envelope, node_id: str) -> dict[str, float]:
    current = _parent_id_of(envelope, node_id)
    while current:
        node = find_node(envelope, current)
        if node:
            c = _first_solid_color(node.get("fills")) or _first_solid_color(node.get("backgrounds"))
            if c:
                return c
        current = _parent_id_of(envelope, current)
    return DEFAULT_BG


def _text_foreground_color(node: TreeNode) -> dict[str, float]:
    c = _first_solid_color(node.get("fills"))
    if c:
        return c
    for seg in node.get("styledSegments") or []:
        c = _first_solid_color((seg.get("style") or {}).get("fills"))
        if c:
            return c
    return DEFAULT_FG


def _effective_font_size(node: TreeNode, envelope: Envelope) -> float | None:
    fs = node.get("fontSize")
    if isinstance(fs, (int, float)):
        return float(fs)
    tsid = node.get("textStyleId")
    if tsid:
        for style in envelope.get("textStyles") or []:
            if style.get("id") == tsid and isinstance(style.get("fontSize"), (int, float)):
                return float(style["fontSize"])
    for seg in node.get("styledSegments") or []:
        sfs = (seg.get("style") or {}).get("fontSize")
        if isinstance(sfs, (int, float)):
            return float(sfs)
    return None


def _font_signature(node: TreeNode, envelope: Envelope) -> str:
    size = _effective_font_size(node, envelope)
    style_name = None
    tsid = node.get("textStyleId")
    if tsid:
        for style in envelope.get("textStyles") or []:
            if style.get("id") == tsid:
                style_name = style.get("name")
                break
    family = (node.get("fontName") or {}).get("family") or style_name or "default"
    return f"{family}:{size if size is not None else 'inherit'}"


def _text_nodes_in_content(after: Envelope, content_ids: set[str]) -> list[TreeNode]:
    nodes: list[TreeNode] = []
    for nid in content_ids:
        node = find_node(after, nid)
        if node and node.get("type") == "TEXT":
            nodes.append(node)
    return nodes


def run_heuristics(_before: Envelope, after: Envelope, graph: EditGraph) -> list[SubCheckResult]:
    content_ids = collect_content_node_ids(graph, after)
    if not content_ids:
        reason = {"reason": "no_content_changes"}
        return [
            _heur_result("heuristics.contrast", 1.0, False, reason),
            _heur_result("heuristics.font_count", 1.0, False, reason),
            _heur_result("heuristics.readable_font_size", 1.0, False, reason),
        ]

    texts = _text_nodes_in_content(after, content_ids)

    if not texts:
        contrast = _heur_result("heuristics.contrast", 1.0, False, {"reason": "no_text_in_content"})
        font_count = _heur_result("heuristics.font_count", 1.0, False, {"reason": "no_text_in_content"})
        readable = _heur_result("heuristics.readable_font_size", 1.0, False, {"reason": "no_text_in_content"})
    else:
        ratios = [
            _contrast_ratio(_text_foreground_color(t), _resolve_background_color(after, t["id"]))
            for t in texts
        ]
        mean_ratio = sum(ratios) / len(ratios)
        contrast = _heur_result(
            "heuristics.contrast",
            min(1.0, mean_ratio / WCAG_AA_NORMAL),
            True,
            {
                "texts_checked": len(texts),
                "mean_contrast_ratio": mean_ratio,
                "min_ratio": min(ratios),
            },
        )

        signatures = {_font_signature(t, after) for t in texts}
        count = len(signatures)
        fc_score = (
            1.0
            if count <= MAX_DISTINCT_FONTS
            else max(0.0, 1.0 - (count - MAX_DISTINCT_FONTS) / MAX_DISTINCT_FONTS)
        )
        font_count = _heur_result(
            "heuristics.font_count",
            fc_score,
            True,
            {"distinct_fonts": count, "max_recommended": MAX_DISTINCT_FONTS},
        )

        violations = 0
        sizes: list[float] = []
        for t in texts:
            size = _effective_font_size(t, after)
            if size is None:
                continue
            sizes.append(size)
            if size < MIN_READABLE_FONT_SIZE:
                violations += 1
        if not sizes:
            readable = _heur_result(
                "heuristics.readable_font_size", 1.0, False, {"reason": "no_resolved_font_sizes"}
            )
        else:
            readable = _heur_result(
                "heuristics.readable_font_size",
                1.0 - violations / len(sizes),
                True,
                {
                    "violations": violations,
                    "sizes_checked": len(sizes),
                    "min_size": min(sizes),
                },
            )

    return [contrast, font_count, readable]
